### PR TITLE
Add `image_pull_policy` option to `KubernetesRun` config

### PR DESCRIPTION
**Pull request by zschumacher**
## Summary

Adds an `image_pull_policy` to `KubernetesRun` config object

## Changes

- Adds an `image_pull_policy` kwarg to `KubernetesRun` config object
- Adds imagePullPolicy to job template if `image_pull_policy` is `Always`, `Never` or `IfNotPresent`

## Importance

This is a common gotcha for Kubernetes jobs, and is a result of the discussion on #4445 

Closes #4445 

## Checklist

This PR:

- [x] adds new tests (if appropriate)
- [x] adds a change file in the `changes/` directory (if appropriate)
- [x] updates docstrings for any new functions or function arguments, including `docs/outline.toml` for API reference docs (if appropriate)

### CODE REVIEW

**Review by slightstone — DISMISSED** [review]

File: src/prefect/serialization/run_config.py [hunk 1]
```diff
@@ -32,6 +32,7 @@ class Meta:
     memory_request = fields.String(allow_none=True)
     service_account_name = fields.String(allow_none=True)
     image_pull_secrets = fields.List(fields.String(), allow_none=True)
+    always_pull_new_image = fields.String(allow_none=False)
```
Comment (zangell44) [inline src/prefect/serialization/run_config.py:35]: This should be `fields.Boolean`, I think this is why "False == False" wasn't working in your test case

File: tests/serialization/test_run_configs.py [hunk 2]
```diff
@@ -62,9 +63,14 @@ def test_serialize_kubernetes_run(config):
         "memory_request",
         "service_account_name",
         "image_pull_secrets",
+        "always_pull_new_image",
     ]
     for field in fields:
-        assert getattr(config, field) == getattr(config2, field)
```
Comment (zangell44) [inline tests/serialization/test_run_configs.py:67]: I think the serialization field type is causing issues here, we should not have to change the test logic here

File: tests/agent/test_k8s_agent.py [hunk 3]
```diff
@@ -1262,10 +1262,19 @@ def test_generate_job_spec_uses_job_template_provided_in_run_config(self):
...
         labels["TEST"] = "VALUE"
-        flow_run = self.build_flow_run(KubernetesRun(job_template=template))
+        config = KubernetesRun(job_template=template, always_pull_new_image=True)
+
+        flow_run = self.build_flow_run(config)
         job = self.agent.generate_job_spec(flow_run)
         assert job["metadata"]["labels"]["TEST"] == "VALUE"
 
+        # test always pull new image kwarg works
+        print(type(job))
```
Comment (zangell44) [inline tests/agent/test_k8s_agent.py:1272]: We can take this debugging statement out

File: tests/agent/test_k8s_agent.py [hunk 4]
```diff
@@ -1262,10 +1262,19 @@ def test_generate_job_spec_uses_job_template_provided_in_run_config(self):
         template = self.read_default_template()
         labels = template.setdefault("metadata", {}).setdefault("labels", {})
         labels["TEST"] = "VALUE"
-        flow_run = self.build_flow_run(KubernetesRun(job_template=template))
```
Comment (zangell44) [inline tests/agent/test_k8s_agent.py:1265]: I think this belongs in a separate test case, something like

```python
@pytest.mark.parameterize(
    ('always_pull', 'expected_pull_policy')
    [
        (True, 'Always'),
        (False, '<default pull policy here>')
    ]
def test_generate_job_spec_sets_image_pull_policy_from_run_config(self, always_pull, expected_pull_policy):
    template = self.read_default_template()
    config = KubernetesRun(job_template=template, always_pull_new_image=always_pull)
    flow_run = self.build_flow_run(config)
    job = self.agent.generate_job_spec(flow_run)
    assert (
            job["spec"]["template"]["spec"]["containers"][0]["imagePullPolicy"]
            == expected_pull_policy
        )
```

File: src/prefect/run_configs/kubernetes.py [hunk 5]
```diff
@@ -38,6 +39,10 @@ class KubernetesRun(RunConfig):
         - labels (Iterable[str], optional): an iterable of labels to apply to this
             run config. Labels are string identifiers used by Prefect Agents
             for selecting valid flow runs when polling for work
+        - always_pull_new_image (bool): a boolean flag for telling kubernetes to
```
Comment (zangell44) [inline src/prefect/run_configs/kubernetes.py:42]: nit: we should add "Defaults to False" somewhere in the docstring

File: src/prefect/serialization/__init__.py [hunk 6]
```diff
@@ -6,3 +6,4 @@
 import prefect.serialization.state
 import prefect.serialization.environment
 import prefect.serialization.storage
+import prefect.serialization.run_config
```
Comment (zangell44) [inline src/prefect/serialization/__init__.py:9]: do we need to add this import?

File: docs/orchestration/flow_config/run_configs.md [hunk 7]
```diff
@@ -189,6 +189,14 @@ Use a custom Kubernetes Job spec for this flow, stored in S3:
 flow.run_config = KubernetesRun(job_template_path="s3://bucket/path/to/spec.yaml")
 ```
 
+Always pull the newest image for the job run.  This should be used when using a custom docker image 
+with a tag other than `:latest`.  Doing so will set the kubernetes [imagePullPolicy](https://kubernetes.io/docs/concepts/containers/images/#updating-images)
+to `Always` on the job run.
+
+```python
+flow.run_config = KubernetesRun(always_pull_latest_image=True)
```
Comment (zangell44) [inline docs/orchestration/flow_config/run_configs.md:197]: The argument in the code is `always_pull_new_image`. I actually like `always_pull_latest_image` better. I would vote to update to `always_pull_latest_image` everywhere.

**Review by zangell44 — CHANGES_REQUESTED** [review]
Thanks for putting this together @zschumacher, this will be super valuable! I left a few comments with requests for changes

File: docs/orchestration/flow_config/run_configs.md (hunk 7 above)
Comment (jcrist) [reply docs/orchestration/flow_config/run_configs.md:197]: Why not just expose `image_pull_policy` directly? Having a nonstandard kwarg will confuse those that are used to k8s, and we already make use of k8s kwarg names elsewhere.

Comment (zangell44) [reply docs/orchestration/flow_config/run_configs.md:197]: yeah fair point I like that better, +1 for exposing `image_pull_policy` directly

Comment (zschumacher) [reply docs/orchestration/flow_config/run_configs.md:197]: yep, makes sense to me.  Do you have a preference on how that is exposed based on the rest of the repo? e.g.

1. type hint `Literal[Always, w/e]`
2. enum
3. just a string

Comment (jcrist) [reply docs/orchestration/flow_config/run_configs.md:197]: Just a string makes sense to me. We can check in the `KubernetesRun` constructor that the value is valid.

**Review by zangell44 — DISMISSED** [review]
This looks great! Thanks @zschumacher!

File: src/prefect/run_configs/kubernetes.py [hunk 8]
```diff
@@ -108,6 +121,16 @@ def __init__(
         if image_pull_secrets is not None:
             image_pull_secrets = list(image_pull_secrets)
 
+        image_pull_policies = {"Always", "IfNotPresent", "Never"}
+        if (
+            image_pull_policy is not None
+            and image_pull_policy not in image_pull_policies
+        ):
+            raise ValueError(
+                f"Invalid image_pull_policy {image_pull_policy}.  "
```
Comment (madkinsz) [inline src/prefect/run_configs/kubernetes.py:130]:
```suggestion
                f"Invalid image_pull_policy {image_pull_policy!r}.  "
```

For displaying string values like this, it's nice to get the repr which will put quotes so

> Invalid image_pull_policy Foo bar

vs

> Invalid image_pull_policy 'Foo bar'

File: src/prefect/run_configs/kubernetes.py [hunk 9]
```diff
@@ -108,6 +121,16 @@ def __init__(
...
+                "Expected `Always`, `IfNotPresent`, or `Never`"
```
Comment (madkinsz) [inline src/prefect/run_configs/kubernetes.py:131]:
```suggestion
                "Expected 'Always', 'IfNotPresent', or 'Never'"
```
We expect string values whereas ` implies an object to me

**Review by madkinsz — CHANGES_REQUESTED** [review]
This looks great! Just a really minor change on the error message

**Pull request merged**